Handle zap_add() failures in "casesensitivity=mixed" mode.

sanjeevbagewadi opened the pull request:
With "casesensitivity=mixed", zap_add() could fail when the number of
files/directories with the same name (varying in case) exceed the
capacity of the leaf node of a Fatzap. This results in a ASSERT()
failure as zfs_link_create() does not expect zap_add() to fail. The fix
is to handle these failures and rollback the transactions.

Signed-off-by: Sanjeev Bagewadi <sanjeev.bagewadi@gmail.com>

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Description
<!--- Describe your changes in detail -->
For a dataset with "casesensitivity=mixed", when a large number of files/directories
with same name (varying only in case e.g: ABCD, ABCd, ABcD and so on) are created
zap_add() could fail. With mixed mode zap_add() normalises the names before the hash
is computed. And all the names would generate the same hash and land in the same leaf.
When the number of entries exceed the capacity of the leaf-block, zap_add() tries to split
the leaf-block which fails as well and zap_add() fails. This trips an ASSERT in zfs_link_create()
as it does not expect zap_add() to fail.

The fix does the following :
- fzap_add_cd() : Handle the case when zap_expand_leaf() fails with ENOSPC and bailout
   without calling zap_put_leaf_maybe_grow_ptrtbl(). 
- zap_add_impl() : When adding to a micro-zap check if the total number of entries
  with colliding/same hash value can fit into fatzap-leaf-block. This is important because, if/when
  the microzap needs to be upgraded to fatzap, all the entries with the same hash would need to
  fit into the same leaf-block (16K). If the number of such entries donot fit fail the zap_add().
  
   The routine mze_canfit_fzap_leaf() today assumes the MZAP_NAME_LEN for every entry.
   This is erring on the safer side but, ends up accommodating lesser number (127) of entries
    with same hash value in microzap. We could find out the size of name of every mze and that
    would be accurate. But, it is expensive to compute the length every time. Alternatively, we
    could compute the length of each entry and cache it. I felt that the amount of code needed
    for this is not worth the gain. I am open to changing it if necessary.

- zfs_link_create() : Move the call to zap_add() to the beginning and in case of a failure
  return. This ensures that we can bailout easily before making any other modifications to
  the parent-zap or the child-dnode. Keeps the code simpler.
- ZPL interfaces (zfs_create(), zfs_mkdir(), zfs_symlink()) : Handle the failure of zfs_link_create()
  and rollback the operation.

With these changes a call to create a file could fail with ENOSPC. Not the best error value.
But, this is the closest I found. Any alternate suggestions are welcome.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

With "casesensitivity=mixed" it is easy to panic the node with a simple test case
as described in https://github.com/zfsonlinux/zfs/issues/7011

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
The following tests were run : 
- zfs-testsuite
- ztest
- Unit-test described in the #7011 

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

ahrens commented on module/zfs/zap_micro.c:
```diff
@@ -363,6 +363,43 @@ mze_find_unused_cd(zap_t *zap, uint64_t hash)
+		/*
+		 * Each mzap entry requires at max : 4 chunks
+		 * 3 chunks for names + 1 chunk for value.
+		 */
```
This comment seems out of place; seems like it is relevant to MZAP_ENT_CHUNKS.

sanjeevbagewadi replied:
You are right. Will fix it.

ahrens commented on module/zfs/zap_micro.c:
```diff
@@ -1202,6 +1239,10 @@ zap_add_impl(zap_t *zap, const char *key,
 		} else {
+			if (!mze_canfit_fzap_leaf(zn, zn->zn_hash)) {
+				zap_unlockdir(zap, tag);
+				return (ENOSPC);
```
Did you consider upgrading to a fat zap in this case?  That would mitigate the slightly-smaller-than-necessary limit on collisions in the microzap due to your overestimation of the name lengths.

sanjeevbagewadi replied:
I think that makes sense. Will fix it.

ahrens commented on module/zfs/zfs_dir.c:
```diff
@@ -776,6 +776,28 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
+		 * transaction which will rollback the SA updates done above.
+		 */
+		mutex_exit(&zp->z_lock);
+		return (error);
```
need to update the comment on line 745.

sanjeevbagewadi replied:
Will do.

ahrens commented on module/zfs/zfs_vnops.c:
```diff
@@ -1445,10 +1446,25 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
+			 */
+			zp->z_unlinked = 1;
+			drop_nlink(ZTOI(zp));
+			zfs_unlinked_add(zp, tx);
```
adding this to the unlinked set seems like a roundabout way of removing it.  Can we just call zfs_znode_delete()?  I know zfs_inactive->zfs_zinactive->zfs_rmnode has some additional logic, but I don't think that it applies in this case.  E.g. this znode can not have an xattrdir at this point (I think).

The same applies to symlink and mkdir.

sanjeevbagewadi replied:
Will do that.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -1429,6 +1429,7 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
 			dmu_tx_hold_write(tx, DMU_NEW_OBJECT,
 			    0, acl_ids.z_aclp->z_acl_bytes);
 		}
+		dmu_tx_hold_zap(tx, zfsvfs->z_unlinkedobj, FALSE, NULL);
```
This shouldn't be needed anymore after switching to `zfs_znode_delete`.  Here and elsewhere in this patch.

sanjeevbagewadi replied:
Makes sense. Will take care of it.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -2039,13 +2054,22 @@ zfs_mkdir(struct inode *dip, char *dirname, vattr_t *vap, struct inode **ipp,
+		dmu_tx_commit(tx);
+		iput(ZTOI(zp));
+		ZFS_EXIT(zfsvfs);
+		return (error);
```
A similar thing could be done here with an `out` label, it might make the code easier to read.

sanjeevbagewadi commented on module/zfs/zfs_dir.c:
```diff
@@ -776,6 +780,28 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
+	/*
+	 * zap_add could fail to add the entry if it exceeds the capacity of the
+	 * leaf-block and zap_leaf_split() failed to help. In such a case bail
+	 * out.
```
Done

sanjeevbagewadi commented on module/zfs/zfs_dir.c:
```diff
@@ -776,6 +780,28 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
+	 * out.
+	 */
+	if (error != 0) {
+		if (!(flag & ZRENAMING) && !(flag &ZNEW)) {
```
Done

sanjeevbagewadi commented on module/zfs/zfs_vnops.c:
```diff
@@ -1445,10 +1446,23 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
+			iput(ZTOI(zp));
+			ZFS_EXIT(zfsvfs);
+			return (error);
+		}
```
Let me give that shot.

sanjeevbagewadi commented on module/zfs/zfs_vnops.c:
```diff
@@ -3685,6 +3709,9 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 			}
+		} else {
+			VERIFY3U(zfs_link_create(tdl, tzp, tx, ZRENAMING),
+			    ==, 0);
```
Oops.. will fix it.

sanjeevbagewadi commented:
Thank you Brian for the review comments. I am working on them and the test case.

behlendorf commented:
@sanjeevbagewadi in or to make reviewing this easier could you please squash your commits together and rebase the entire PR against master.  You'll need to force update the branch on Github which is fine.

ahrens commented on module/zfs/zfs_vnops.c:
```diff
@@ -3856,25 +3878,34 @@ zfs_symlink(struct inode *dip, char *name, vattr_t *vap, char *link,
+	error = zfs_link_create(dl, zp, tx, ZNEW);
+	if (error != 0) {
+		zfs_znode_delete(zp, tx);
+		goto out;
```
you don't need the `goto` here

sanjeevbagewadi replied:
Oops... makes sense.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+#!/bin/ksh -p
```
You're going to need to add to add this test case to the `casenorm` section of the `tests/runfiles/linux.run`.  This is why it wasn't run by the automated testing.

sanjeevbagewadi replied:
Ah ! I missed that.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+# when multiple files which the same name (differing only in case) created, the
+# number of files is limited to what can fit in a fatzap leaf-block. And beyond
+# that, it fails with ENOSPC. Ensure that the create fails gracefully and does
+#  not hit and assert and panic.
```
[nit] extra leading whitespace
s/not hit and assert and panic./not hit an assert and panic./

sanjeevbagewadi replied:
Done.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+
+# DESCRIPTION:
+# For the filesystem with casesensitivity=mixed, normalization=none,
+# when multiple files which the same name (differing only in case) created, the
```
s/which/with/
s/created/are created/

sanjeevbagewadi replied:
Done.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+
+names='{a,A}{b,B}{c,C}{d,D}{e,E}{f,F}{g,G}{h,H}{i,I}{j,J}{k,K}{l,L}'
+for name in $names; do
+        create_file $name
```
Might be Github but it looks like there's an indent issue here.

sanjeevbagewadi replied:
There was an extra space. Deleted it.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+        create_file $name
+	ret=$?
+	log_note "Created file $name status $ret"
+	if (($ret != 0)); then
```
It would be nice if we could do this in a way where the test checks for ENOSPC and just any error.  That way we know we're failing for the right reason, that said I don't think it's critical.

sanjeevbagewadi replied:
will try to address this. The exit code does not indicate ENOSPC. Will need to rely on the error message.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+	fi
+done
+
+destroy_testfs
```
This could be removed since the onexit handler always runs.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+
+destroy_testfs
+
+log_pass "CM-not-UN FS: lookup succeeds only if using exact name form"
```
Please update the message here to reflect what's being tested.

sanjeevbagewadi replied:
oops.. slip up. Fixed it.

behlendorf commented on module/zfs/zap.c:
```diff
@@ -850,10 +854,29 @@ fzap_add_cd(zap_name_t *zn,
+		 * trying to expand again
+		 */
+		if (expand_retries > MAX_EXPAND_RETRIES && prev_l == l) {
+			err = ENOSPC;
```
You should use `err = SET_ERROR(ENOSPC);` here, similar to what's done a few lines up for `EEXIST`.

sanjeevbagewadi replied:
Done

behlendorf commented on module/zfs/zap_micro.c:
```diff
@@ -363,6 +363,43 @@ mze_find_unused_cd(zap_t *zap, uint64_t hash)
+ * Check if the current entry keeps the colliding entries under the fatzap leaf
+ * size.
+ *
+ * This is important to ensure that the microzap is upgradable to fatzap.
```
[nit] I'm not sure this needs to be its own paragraph, you could include it after the sentence above.

sanjeevbagewadi replied:
Deleting that line as it is no longer correct. As per Matt's suggestion, we now trigger an upgrade to fatzap when the number of chunks exceed the limit.

Without this change we were failing the zap_add() for a smaller number if it was a microzap.
With upgrade, we would first upgrade to fatzap and that would allow one to add more entries with colliding hash. And eventually zap_add() to fatzap would fail when the colliding entry can longer fit in the leaf-block.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -1446,10 +1447,20 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
+		if (error != 0) {
+			/*
+			 * Delete the node created above.
+			 */
```
[nit]  I'd drop the comment here or update it to be clear why `zp` from `zfs_mknode` needs to be destroyed.

sanjeevbagewadi replied:
Have reworded the comment. If does not serve the purpose, I am fine with deleting it.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -3686,6 +3706,8 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 				    ZRENAMING, NULL), ==, 0);
 			}
+		} else {
+			VERIFY0(zfs_link_create(tdl, tzp, tx, 0));
```
Why is this only needed now?  This looks like it may address a potential existing issue.

sanjeevbagewadi replied:
You are right, this seems like an existing issue. But, it is mostly a remote/unlikely case for the following reasons :
- We are deleting an entry on line:3677 and adding the same entry back on line:3680.
   Hence, i don't think it would fail.  Also, since this thread holds, the zfs_dirlock(), no other
   thread can race us.

However, as a safety measure, decided to add this code here.

tuxoko replied:
@sanjeevbagewadi 
I'll copy me gerrit comment here:
> You haven't add 'if (tzp)'
>
>Also, the zfs_link_destroy above will set z_unlinked to true, which will cause this to fail. And it will call zfs_unlinked_add. You'll need to revert those changes.

I don't feel we need to solve this in this patch, since this path is probably not very likely, and it's an existing issue. Maybe we should just have a warning here, and solve it later.

sanjeevbagewadi replied:
Agreed. I had overlooked the fact the zfs_link_create() will not solve the problem. For now I will add an ASSERT() here. And we can take that up separately.

behlendorf commented on module/zfs/zfs_vnops.c:
```diff
@@ -3856,25 +3878,34 @@ zfs_symlink(struct inode *dip, char *name, vattr_t *vap, char *link,
+		zfs_inode_update(zp);
+	}
 
+out:
```
Or the `out:`.

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,56 @@
+# STRATEGY:
+# 1. Create files with same name but varying in case.
+#    E.g. 'abcdefghijklmnop', 'Abcdefghijklmnop', 'ABcdefghijklmnop' etc.
+#    The create would fail with ENOSPC.
```
Since this change touches several system calls we should extent the test case to verify they behave as expected too.  Specifically let's check rename, symlinks, and mkdir.

tuxoko commented on module/zfs/zfs_vnops.c:
```diff
@@ -1446,10 +1447,21 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
+			 * Since, we failed to add the directory entry for it,
+			 * delete the newly created dnode.
+			 */
+			zfs_znode_delete(zp, tx);
```
@sanjeevbagewadi
I'll copy me gerrit comment here:
>iput_final won't evict the inode unless you drop the link count to zero or call remove_inode_hash to unhash it. But if you just drop the link count, you will still have a window that might have obj id reuse issue I mentioned above.
>
>So the best way I think is to call remove_inode_hash.

sanjeevbagewadi replied:
makes sense. Will add remove_inode_hash().

sanjeevbagewadi commented:
@behlendorf , I have updated the test case to cover the different object types (files, dir, symlink, hardlinks) and for each object type exercise the create and rename operation. Kindly take a look.

@ahrens , for the corner case in zfs_rename() for now, I have added an assert() there. If you agree we could take that up as a separate case.

@tuxoko , I have now added remove_inode_hash().

behlendorf commented on tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh:
```diff
@@ -0,0 +1,137 @@
+#    failed to add in step 1 above.
+#    This should fail as well.
+
+set -x
```
nit: Let's drop the `set -x` for the final version.
